File: scheduler.py
#!/usr/bin/env python3
"""
Scheduler for automated trending data fetching and report generation
"""
import logging
import schedule
import time
from datetime import datetime
from src.config.settings import settings
from src.database.base import SessionLocal
from src.summarize import ProjectSummarizer
from src.generate import ReportGenerator

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


def daily_job():
    """Single daily job - fetch, summarize, and generate report"""
    logger.info("Starting daily GitHub trending job...")
    db = SessionLocal()

    try:
        # 1. Fetch trending repositories
        logger.info("Fetching trending repositories...")
        from src.fetch_data import TrendingScraper
        scraper = TrendingScraper(db)
        count = scraper.fetch_and_save(since="daily")
        logger.info(f"Fetched {count} trending repositories")

        # 2. Generate AI summaries (optional, limited to save costs)
        logger.info("Generating summaries for new projects...")
        summarizer = ProjectSummarizer(db)
        summary_count = summarizer.batch_summarize(limit=5)  # Only 5 to save costs
        logger.info(f"Generated {summary_count} summaries")

        # 3. Generate daily report
        logger.info("Generating daily report...")
        report_gen = ReportGenerator(db)
        filepath = report_gen.generate_and_save()
        logger.info(f"Report generated: {filepath}")

        logger.info("Daily job completed successfully!")

    except Exception as e:
        logger.error(f"Error in daily job: {e}", exc_info=True)
    finally:
        db.close()


def main():
    """Main scheduler loop"""
    logger.info("GitHub Trending Scheduler started")

    # Schedule single daily job at 10:00 AM
    schedule.every().day.at("10:00").do(daily_job)

    logger.info("Job scheduled: Daily update at 10:00 AM")

    # Run immediately on start (optional)
    # daily_fetch_job()

    # Keep running
    while True:
        schedule.run_pending()
        time.sleep(60)  # Check every minute


if __name__ == "__main__":
    main()

File: src/scheduler.ts
import 'reflect-metadata';
import * as cron from 'node-cron';
import * as dotenv from 'dotenv';
import { initializeDatabase, AppDataSource } from './database/data-source';
import { Project } from './database/models/Project';
import { TrendingSnapshot } from './database/models/TrendingSnapshot';
import { Summary } from './database/models/Summary';
import { TrendingScraper } from './fetch/TrendingScraper';
import { ProjectSummarizer } from './summarize/ProjectSummarizer';

dotenv.config();

async function fetchTrendingData() {
  console.log('Starting scheduled trending fetch...');
  try {
    const projectRepo = AppDataSource.getRepository(Project);
    const snapshotRepo = AppDataSource.getRepository(TrendingSnapshot);

    const scraper = new TrendingScraper(projectRepo, snapshotRepo);
    const count = await scraper.fetchAndSave(undefined, 'daily');

    console.log(`✓ Successfully fetched ${count} trending repositories`);
  } catch (error) {
    console.error('Error fetching trending data:', error);
  }
}

async function generateSummaries() {
  console.log('Starting scheduled summary generation...');
  try {
    const projectRepo = AppDataSource.getRepository(Project);
    const summaryRepo = AppDataSource.getRepository(Summary);

    const summarizer = new ProjectSummarizer(projectRepo, summaryRepo);
    const summaries = await summarizer.summarizeTrendingProjects(5);

    console.log(`✓ Successfully generated ${summaries.length} summaries`);
  } catch (error) {
    console.error('Error generating summaries:', error);
  }
}

async function main() {
  try {
    // Initialize database
    await initializeDatabase();
    console.log('✓ Database initialized');

    // Schedule tasks
    // Fetch trending data every day at 9:00 AM
    cron.schedule('0 9 * * *', fetchTrendingData, {
      scheduled: true,
      timezone: 'UTC',
    });
    console.log('✓ Scheduled daily trending fetch at 9:00 AM UTC');

    // Generate summaries every day at 10:00 AM
    cron.schedule('0 10 * * *', generateSummaries, {
      scheduled: true,
      timezone: 'UTC',
    });
    console.log('✓ Scheduled daily summary generation at 10:00 AM UTC');

    // Run once immediately on startup
    console.log('\nRunning initial data fetch...');
    await fetchTrendingData();

    console.log('\nScheduler is running. Press Ctrl+C to exit.');

    // Keep the process alive
    process.on('SIGINT', async () => {
      console.log('\nShutting down scheduler...');
      process.exit(0);
    });

    process.on('SIGTERM', async () => {
      console.log('\nShutting down scheduler...');
      process.exit(0);
    });
  } catch (error) {
    console.error('Failed to start scheduler:', error);
    process.exit(1);
  }
}

main();
